### PR TITLE
Feature/140 self update

### LINKED ISSUE #140: Self update
Add a feature to project keeper to update itself to the latest available version.

### PR DESCRIPTION
**Pull request by ckunki**
fixes #140

### CODE REVIEW

File: project-keeper/src/main/java/com/exasol/projectkeeper/mavenrepo/Version.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Arrays;
+import java.util.regex.Pattern;
+
+import lombok.EqualsAndHashCode;
+
+/**
+ * Parse a version number and implement Comparable for that.
+ *
+ * <p>
+ * Compared to maven-artifact-3.8.5.jar, package org.apache.maven.artifact.versioning.ComparableVersion this class
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/mavenrepo/Version.java:12]:
```suggestion
 * Compared to class {@code org.apache.maven.artifact.versioning.ComparableVersion} from Maven this class
```

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import com.exasol.projectkeeper.mavenrepo.MavenRepository.JsonContentException;
+import com.exasol.projectkeeper.mavenrepo.Version.UnsupportedVersionFormatException;
+
+import jakarta.json.*;
+
+//[utest->dsn~verify-own-version~1]
+class LatestVersionTest {
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java:22]: - Split into MavenRepositoryTest & VersionTest
- Add tests for hashcode & equals, see https://jqno.nl/equalsverifier/

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @ParameterizedTest(name = "Version(\"{0}\")")
+    @CsvSource(value = { "0", "0.1", "02.03", "000.111.222" })
+    void validVersionStrings(final String version) throws UnsupportedVersionFormatException {
+        final Version testee = new Version(version);
+        assertThat(testee, notNullValue());
+        assertThat(testee.toString(), equalTo(version));
+    }
+
+    @ParameterizedTest(name = "Version(\"{0}\")")
+    @CsvSource(value = { "''", "aa", "1 2", ".1", "000.111.222." })
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java:35]:
```suggestion
    @CsvSource(value = { "''", "aa", "1.2.c", "1 2", ".1", "000.111.222." })
```

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    @Test
+    void invalidJson_ThrowsException() throws IOException {
+        assertThrows(JsonException.class, () -> json("{"));
+    }
+
+    private JsonObject json(final String content) {
+        try (JsonReader jr = Json.createReader(new StringReader(content))) {
+            return jr.readObject();
+        }
+    }
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java:71]:
```suggestion
```

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        assertThat(url, startsWith(MavenRepository.DEFAULT_REPOSITORY_URL));
+        assertThat(url, containsString(MavenRepository.GROUP_ID));
+        assertThat(url, containsString(MavenRepository.ARTIFACT_PREFIX + "-cli"));
+    }
+
+    @Test
+    void jsonWithoutVersionInformation_ThrowsException() throws IOException {
+        final JsonObject json = json("{\"response\": {\"docs\": [{\"other\": \"2.4.6\"}]}}");
+        assertThrows(JsonContentException.class, () -> MavenRepository.getLatestVersion(json));
+    }
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java:87]:
```suggestion
    @Test
    void emptyJson_ThrowsException() throws IOException {
        final JsonObject json = json("{}");
        assertThrows(JsonContentException.class, () -> MavenRepository.getLatestVersion(json));
    }
    
    @Test
    void jsonWithoutVersionInformation_ThrowsException() throws IOException {
        final JsonObject json = json("{\"response\": {\"docs\": [{\"other\": \"2.4.6\"}]}}");
        assertThrows(JsonContentException.class, () -> MavenRepository.getLatestVersion(json));
    }
```

File: project-keeper/src/test/java/com/exasol/projectkeeper/validators/OwnVersionValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        final SimpleValidationFinding finding = (SimpleValidationFinding) testee.validate().get(0);
+        assertThat(finding.getMessage(), matchesRegex(OUTDATED));
+        assertThat(finding.hasFix(), is(false));
+    }
+
+    // private helpers
+
+    private OwnVersionValidator testee(final String currentVersion) {
+        return testee(currentVersion, new MavenRepository("https://localhost/unknown/end/point"));
+    }
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/validators/OwnVersionValidatorTest.java:84]: - Rename method to "testeeWithInvalidMavenRepo".
- Check if argument is necessary or could be null

File: project-keeper/src/test/resources/com/exasol/projectkeeper/mavenrepo/json.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{"response": {
```
Comment (kaklakariada) [inline project-keeper/src/test/resources/com/exasol/projectkeeper/mavenrepo/json.json:1]: Rename file to e.g. maven-central-version-response.json

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    public static OwnVersionValidator forMavenPlugin(final String currentVersion, final Updater updater) {
+        return new OwnVersionValidator(currentVersion, MavenRepository.mavenPlugin(), updater);
+    }
+
+    /**
+     * @param currentVersion current version of PK in order to validate if there is an update available.
+     * @return instance of {@link OwnVersionValidator} without the ability to perform a self-update
+     */
+    // fix is only possible if using PK from maven pom
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:41]: Add to javadoc comment

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    private static List<ValidationFinding> findings(final Fix fix, final String message) {
+        final Builder builder = SimpleValidationFinding.withMessage(message).optional(true);
+        if (fix != null) {
+            builder.andFix(fix);
+        }
+        return List.of(builder.build());
+    }
+
+    // instance fields and methods
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:54]:
```suggestion
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        } catch (final ValidationException e) {
+            return findings(null, e.getMessage());
+        }
+    }
+
+    private Version parseVersion(final String version, final ValidationException exception) throws ValidationException {
+        try {
+            return new Version(version);
+        } catch (final UnsupportedVersionFormatException e) {
+            throw exception;
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:95]:
```suggestion
            exception.setCause(e);
            throw exception;
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    private Version getLatestVersion(final MavenRepository repo) throws ValidationException {
+        try {
+            final String versionString = repo.getLatestVersion();
+            return parseVersion(versionString, new ValidationException(ExaError.messageBuilder("W-PK-CORE-154") //
+                    .message("Could not detect latest available version of project-keeper.") //
+                    .message(" Unsupported format of latest version from Maven repository: {{version}}.", versionString) //
+                    .toString()));
+        } catch (final IOException | JsonContentException e) {
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:106]:
```suggestion
        } catch (final IOException | JsonContentException exception) {
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            throw new ValidationException(ExaError.messageBuilder("W-PK-CORE-155") //
+                    .message("Could not detect latest available version of project-keeper.") //
+                    .message(" {{message|uq}}.", e.getMessage()) //
+                    .mitigation("Please check network connection and response from {{url}}", repo.getUrl()) //
+                    .toString());
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:111]:
```suggestion
            throw new ValidationException(ExaError.messageBuilder("W-PK-CORE-155") //
                    .message("Could not detect latest available version of project-keeper.") //
                    .message(" {{message|uq}}.", e.getMessage()) //
                    .mitigation("Please check network connection and response from {{url}}", repo.getUrl()) //
                    .toString(), exception);
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    private static class ValidationException extends Exception {
+        private static final long serialVersionUID = 1L;
+
+        public ValidationException(final String message) {
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/OwnVersionValidator.java:118]: Add cause to constructor

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/finding/SimpleValidationFinding.java [hunk 14]
```diff
@@ -1,29 +1,52 @@
...
+     */
+    public static List<SimpleValidationFinding> blockers(final List<SimpleValidationFinding> findings) {
+        return findings.stream() //
+                .filter(not(SimpleValidationFinding::isOptional)) //
+                .collect(Collectors.toList());
+    }
+
     private final String message;
+    @Getter
+    @Accessors(fluent = true)
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/finding/SimpleValidationFinding.java:36]: Check if this makes sense

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/finding/SimpleValidationFinding.java (hunk 14 above)
Comment (ckunki) [reply project-keeper/src/main/java/com/exasol/projectkeeper/validators/finding/SimpleValidationFinding.java:36]: yes makes sense, but is redundant.
> fluent -- If true, accessors will be named after the field and not include a get or setprefix.

Default getter generated by lombok is already "isOptional()".

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java (hunk 2 above)
Comment (ckunki) [reply project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/LatestVersionTest.java:22]: Uh Oh: @EqualsAndHashCode and nl.jqno.equalsverifier.EqualsVerifier ... 
discussion https://github.com/projectlombok/lombok/issues/1872  ended in `suppress(Warning.STRICT_INHERITANCE)` 😦

File: project-keeper/src/test/resources/com/exasol/projectkeeper/mavenrepo/json.json (hunk 7 above)
Comment (ckunki) [reply project-keeper/src/test/resources/com/exasol/projectkeeper/mavenrepo/json.json:1]: Moved file  to `test/resources/simulated-maven-central-version-response.json`

File: project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/VersionTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.exasol.projectkeeper.mavenrepo.Version.UnsupportedVersionFormatException;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+import nl.jqno.equalsverifier.Warning;
+
+//[utest->dsn~verify-own-version~1]
+class VersionTest {
+
+    // test Version
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/mavenrepo/VersionTest.java:19]:
```suggestion
```

File: project-keeper/src/test/java/com/exasol/projectkeeper/validators/dependencies/renderer/DependencyPageRendererTest.java [hunk 16]
```diff
@@ -53,13 +93,43 @@ void testMultiProjectSetup() {
...
+                "", //
+                "[0]: https://example.com/mylib", //
+                "[1]: https://mit.edu")));
+    }
+
+    private final String NEWLINE = System.lineSeparator();
+
+    private String lines(final String... lines) {
+        return Stream.of(lines).collect(Collectors.joining(this.NEWLINE, "", this.NEWLINE));
+    }
```
Comment (kaklakariada) [inline project-keeper/src/test/java/com/exasol/projectkeeper/validators/dependencies/renderer/DependencyPageRendererTest.java:120]: 👍

**Review by kaklakariada — CHANGES_REQUESTED** [review]

**Pull request merged**